### PR TITLE
Ensure JSON representation is compact. #3363

### PR DESCRIPTION
**Pull request by BERRADA-Omar**
# Summary

this PR solves the issues mentionned here #3363

> The following related issues could all be resolved together...
> 
> We should use ensure_ascii=False for more compact text representations.
> We should use separators = (',', ':') for more compact list and object representations.
> We should use allow_nan=False to disallow invalid Infinity and NaN representations.

# Checklist

- [x] I understand that this PR may be closed in case there was no previous discussion. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [x] I've updated the documentation accordingly.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -10,3 +10,4 @@ venv*/
 .python-version
 build/
 dist/
+.vscode/
```
Comment (tomchristie) [inline .gitignore:13]:
```suggestion
```

File: .gitignore [hunk 2]
```diff
@@ -9,4 +9,4 @@ site/
 venv*/
 .python-version
 build/
-dist/
+dist/
```
Comment (tomchristie) [inline .gitignore:12]:
```suggestion
dist/

```

**Review by tomchristie — APPROVED** [review]
Yep, thx. ☺️

File: tests/test_content.py [hunk 3]
```diff
@@ -4,6 +4,7 @@
 import pytest
 
 import httpx
+from httpx._content import encode_json
```
Comment (tomchristie) [inline tests/test_content.py:7]: I've just noticed this accidentally snuck in a private import.

@BERRADA-Omar would you be up for refactoring these tests?

We can test this against public API using `httpx.Request()`. Eg...

```python
data = {...}
req = httpx.Request("POST", "https://www.example.com/", json=data)
assert req.content == ...
assert req.headers = ...
```

**Comment by Kludex** [comment]
This broke the test suite on Starlette. FYI

**Comment by Kludex** [comment]
This broke the test suite on Starlette.

File: tests/test_content.py (hunk 3 above)
Comment (BERRADA-Omar) [reply tests/test_content.py:7]: Hello, i'll make the change.

Comment (BERRADA-Omar) [reply tests/test_content.py:7]: @tomchristie i see that this is already fixed ?

Comment (tomchristie) [reply tests/test_content.py:7]: Yes. Thanks, tho.

**Comment by beliaev-maksim** [comment]
@tomchristie This PR not only adjusts the ASCII representation but also alters the default JSON separators. These changes lead to unnecessary modifications of JSON objects, resulting in altered validation hashes that broke some of our tests.

**Comment by tomchristie** [comment]
Thoughts...

* CHANGELOG entry here [is insufficiently emphasised](https://github.com/encode/httpx/blob/master/CHANGELOG.md#0280-28th-november-2024). Not an API change, but nonetheless changes to representation-on-the-wire *are* likely to negatively impact teams in the short run, as evidenced.
* Unclear how we should improve evaluation process of trade-offs such as this one.

**Comment by beliaev-maksim** [comment]
@tomchristie More detailed CHANGELOG is a "must". While it won’t solve the technical issue, it will simplify the debugging.  

When moving forward a codebase, tough changes might be necessary, that is obvious. However, providing a grace period for users can ease the process. E.g., you could introduce a warning in the codebase and maintain it for two minor versions before enforcing the update.
As one of the options.

**Pull request merged**